course/massexception.php

```diff
@@ -118,16 +118,53 @@
 						$query = "UPDATE imas_assessment_sessions SET scores=:scores,attempts=:attempts,seeds=:seeds,lastanswers=:lastanswers,";
 						$query .= "reattempting=:reattempting WHERE id=:id";
 						$stm = $DBH->prepare($query);
-						$stm->execute(array(':scores'=>$scorelist, ':attempts'=>$attemptslist, ':seeds'=>$seedslist, ':lastanswers'=>$lalist,
-							':reattempting'=>$reattemptinglist, ':id'=>$row[0]));
+						$update = array(':scores'=>$scorelist, ':attempts'=>$attemptslist, ':seeds'=>$seedslist, ':lastanswers'=>$lalist,
+							':reattempting'=>$reattemptinglist, ':id'=>$row[0]);
+						$stm->execute($update);
+
+						$result = TeacherAuditLog::addTracking(
+							$cid,
+							"Clear Scores",
+							$row[0],
+							array(
+								'clear_type'=>'user assessment',
+								'studentid'=>$stu,
+								'old_attempt'=>$row,
+								'new_attempt'=>$update
+							)
+						);
 					}
 
 				} else if (isset($_POST['forceclear'])) {
+					//get old grades
+					$stm = $DBH->prepare("SELECT userid, bestscores FROM imas_assessment_sessions WHERE userid=? AND assessmentid=$aid");

```

Comment:
This and a few lines down have a mix of named placeholders, ? placeholders, and directly injected parameters.